smokedduck/operators.py

```diff
@@ -0,0 +1,356 @@
+
+    def get_child_join_conds(self):
+        return [self.single_op_table_name + ".in_index"]
+
+    def get_out_index(self):
+        return self.single_op_table_name + ".out_index"
+
+
+class TableScan(Op):
+    def __init__(self, query_id, op_id, parent_join_cond):
+        super().__init__(query_id, self.get_name(), op_id, parent_join_cond)
+        self.inner_op_table_name = self.single_op_table_name + "_0"
+
+    def get_name(self):
+        return "SEQ_SCAN"
+
+    def get_from_string(self):
+        if self.is_root:
+            return self.inner_op_table_name + " AS " + self.single_op_table_name
+        else:
+            return "JOIN " + self.inner_op_table_name + " AS " + self.single_op_table_name \
+                + " ON " + self.parent_join_cond + " = " + self.single_op_table_name + ".out_index"
+
+    def get_child_join_conds(self):
+        return [self.single_op_table_name + ".in_index"]
+
+    def get_out_index(self):
+        return self.single_op_table_name + ".out_index"
+
+
+class GroupBy(Op):
+    def __init__(self, query_id, op_id, parent_join_cond):
+        super().__init__(query_id, self.get_name(), op_id, parent_join_cond)
+        self.source = self.single_op_table_name + "_0"
+        self.sink = self.single_op_table_name + "_1"
+        self.combine = self.single_op_table_name + "_2"  # TODO: integrate combine
+        self.finalize = self.single_op_table_name + "_3"
+
+    def get_name(self):
+        return "HASH_GROUP_BY"
+
+    def get_from_string(self):
+        join_sink_and_finalize = " JOIN " + self.finalize + " ON " + self.source + ".in_index = " + self.finalize + ".out_index" + \
+                                 " JOIN " + self.sink + " AS " + self.single_op_table_name + " ON " + self.finalize + ".in_index = " + \
+                                 self.single_op_table_name + ".out_index"
+        if self.is_root:
+            return self.source + join_sink_and_finalize
+        else:
+            return "JOIN " + self.source + " ON " + self.parent_join_cond + " = " + self.source + ".out_index" + join_sink_and_finalize
+
+    def get_child_join_conds(self):
+        return [self.single_op_table_name + ".in_index"]
+
+    def get_out_index(self):
+        return self.source + ".out_index"
+
+
+class PerfectGroupBy(Op):
+    def __init__(self, query_id, op_id, parent_join_cond):
+        super().__init__(query_id, self.get_name(), op_id, parent_join_cond)
+        self.source = self.single_op_table_name + "_0"
+        self.sink = self.single_op_table_name + "_1"
+        self.combine = self.single_op_table_name + "_2"  # TODO: integrate combine
+        self.finalize = self.single_op_table_name + "_3"
+
+    def get_name(self):
+        return "PERFECT_HASH_GROUP_BY"
+
+    def get_from_string(self):
+        join_sink_and_finalize = " JOIN " + self.finalize + " ON " + self.source + ".in_index = " + self.finalize + ".out_index" + \
```

Comment:
this if finalize is included, else:   join_sink_and_finalize = " JOIN " + self.sink + " AS " + self.single_op_table_name +  " ON " + self.source + ".in_index = " + self.single_op_table_name + ".out_index"